### ISSUE TITLE
Warnings after upgrading to otp-24

### ISSUE THREAD

**Issue by LuisaAPF** [fewlinesco/bamboo_smtp#192]
After upgrading to Elixir 1.12, OTP-24, every e-mail sent using bamboo_smtp (v 4.0.1) generates this warning:
`'Authenticity is not established by certificate path validation' Reason: 'Option {verify, verify_peer} and cacertfile/cacerts is missing'`.

It looks like they are related to this change in Erlang: https://github.com/erlang/otp/pull/4805/files.

If these warnings are harmless, as I think they are, it would be nice to have an option to silence them.

**Comment by alli5723** [comment]
Hi @LuisaAPF, we will look at this issue.

**Comment by alappe** [comment]
I already took a shot at adding the needed options and will push the PR shortly after testing it locally.

https://github.com/alappe/bamboo_smtp/tree/allow_for_more_tls_options

**Comment by LuisaAPF** [comment]
Thanks!

**Comment by alli5723** [comment]
@destne do we have a plan for this? 🙂

**Comment by masoudgs** [comment]
Any updates on #193 ? It seems ready to merge.

**Comment by alli5723** [comment]
@msudgh, from my understanding, merging it will not remove the warning yet. 
@alappe indicated that some changes might be required on gen_smtp to accept more options, I am guessing that comment refers to this section https://github.com/gen-smtp/gen_smtp/blob/master/src/smtp_socket.erl#L46.

I'll be spending more time on this today and I will revert as soon as I have an update.

**Comment by masoudgs** [comment]
Thanks @alli5723
You right seem a little refactor needs for `tls` params.
That'd be great if you end-up something today 🙏

**Comment by alappe** [comment]
Sorry for the delay. I will update the PR in the next hours and post my proposed changes to `gen_smtp` here as well for discussion. But I couldn't get a properly verified smtp connection working yet. Maybe someone else finds the solution.

**Comment by masoudgs** [comment]
@alappe Ill test it and btw you can go to `https://mailtrap.io` for setting up a free TEST SMTP Server, then run tests on your end as well.

**Comment by alappe** [comment]
So, I updated the PR (#193) and this is the state of my changes to `gen_smtp`: https://github.com/alappe/gen_smtp/tree/allow_for_more_tls_options

Any help appreciated.

**Comment by MatheusBueno782** [comment]
Hey @alappe you can test it with our `https://github.com/fewlinesco/bamboo_smtp-sandbox`.  I will test on my side too.

**Comment by alli5723** [comment]
@alappe thanks for the change to gen_smtp. While testing and tracking the possibility of adding new TLS options, I noticed that to get those new options passed to the erlang's SSL module, we need to append the options, something like the snippet below.

```erlang
connect(Host, Options) ->
	AddSockOpts = case proplists:get_value(sockopts, Options) of
		undefined -> [];
		Other -> Other
	end,
	AddTLSOpts = case proplists:get_value(tls_options, Options) of
		undefined -> [];
		TLSOpts -> TLSOpts
	end,
	AdditionalOpts = [binary, {packet, line}, {keepalive, true}, {active, false} | AddSockOpts],
	SockOpts = AdditionalOpts ++ AddTLSOpts,
	Proto = case proplists:get_value(ssl, Options) of
		true ->
			ssl;
		_ ->
			tcp
	end,
	Port = case proplists:get_value(port, Options) of
		undefined when Proto =:= ssl ->
			465;
		OPort when is_integer(OPort) ->
			OPort;
		_ ->
			25
	end,
	Timeout = case proplists:get_value(timeout, Options) of
		undefined -> 5000;
		OTimeout -> OTimeout
	end,
	case smtp_socket:connect(Proto, Host, Port, SockOpts, Timeout) of
		{ok, Socket} ->
			case read_possible_multiline_reply(Socket) of
				{ok, <<"220", Banner/binary>>} ->
					{ok, Socket, Host, Banner};
				{ok, <<"4", _Rest/binary>> = Msg} ->
					quit(Socket),
					throw({temporary_failure, Msg});
				{ok, Msg} ->
					quit(Socket),
					throw({permanent_failure, Msg})
			end;
		{error, Reason} ->
			throw({network_failure, {error, Reason}})
	end.
```

The additional lines needed from the above snippet are:
```erlang
AddTLSOpts = case proplists:get_value(tls_options, Options) of
		undefined -> [];
		TLSOpts -> TLSOpts
	end,
AdditionalOpts = [binary, {packet, line}, {keepalive, true}, {active, false} | AddSockOpts],
SockOpts = [AdditionalOpts | AddTLSOpts],
```

However, I am currently getting the following error, I believe it is related to my certificate.
```erlang
{:error,
 {:network_failure, '173.194.73.108',
  {:error,
   {:options,
    {:cacerts, "link_to_cert"}}}}}
```

An additional note to @LuisaAPF, I think it is logical that the warning was added whenever we use SSL with a default option/verify_none looking at [the issue that gave birth](https://github.com/erlang/otp/issues/4800) to the [erlang merge](https://github.com/erlang/otp/pull/4805/files) you referenced.

As you have correctly mentioned the warning is harmless, after making changes to accept more TLS options, you will only stop getting the warnings if you were to update your configurations to include TLS options + a certfile (Otherwise, it will end up as an ~warning~ error). Alternatively, you might want only error log_level in production, so you don't have a noisy log.

What do you think? 🤔

**Comment by LuisaAPF** [comment]
> after making changes to accept more TLS options, you will only stop getting the warnings if you were to update your configurations to include TLS options + a certfile

I guess propagating the changes in `:ssl` is the right way to handle it :+1: . I'll make sure to update my config after the new version is out. Thanks!

**Comment by alli5723** [comment]
@alappe, @LuisaAPF and @msudgh 

I created a PR to append tls_options on gen_smtp, but specifically allowed an extra connection_option_key for `log_level`
https://github.com/gen-smtp/gen_smtp/pull/265/files

I also created a new PR for bamboo-smtp [here](https://github.com/fewlinesco/bamboo_smtp/pull/196). This is to allow accepting the `tls_log_level` configuration option.

**Comment by LuisaAPF** [comment]
Hi @alli5723, just to confirm, for the new tls options to have an effect, we are waiting on https://github.com/alappe/gen_smtp/commit/f8717d9a36cd285d33604a9d9cb62e409107463f or https://github.com/gen-smtp/gen_smtp/pull/274, correct?

**Comment by deepankar-j** [comment]
It looks like https://github.com/gen-smtp/gen_smtp/pull/274 has been merged and is part of the [gen_smtp v1.2.0](https://github.com/gen-smtp/gen_smtp/releases/tag/1.2.0) release.    

@tgautier: It appears there's already an open PR to update to the latest gen_smtp version for the purposes of addressing the `:hut` warning.  Addressing PR https://github.com/fewlinesco/bamboo_smtp/pull/205 would address this issue as well.

**Comment by tgautier** [comment]
Should be fixed by #205.

Feel free to reopen the issue if it's not the case.
